File: chronicle-pc/pipeline/chronicle_pipeline/topics.py
"""Topic notes including dream symbol clustering."""

from __future__ import annotations

import logging
import re
from collections import defaultdict
from pathlib import Path

from .config import ensure_config
from .entries import load_all_entries
from .notes import mirror_note, topic_slug, write_if_changed
from .paths import resolve_chronicle_dir

log = logging.getLogger("chronicle.topics")

WORD_RE = re.compile(r"[a-zA-Z][a-zA-Z0-9']{3,}")
STOP = {
    "that",
    "this",
    "with",
    "from",
    "have",
    "were",
    "been",
    "they",
    "them",
    "then",
    "when",
    "what",
    "your",
    "about",
    "into",
    "just",
    "like",
    "there",
    "their",
    "would",
    "could",
    "should",
    "dream",
    "dreams",
    "dreamt",
    "dreamed",
}


def _render_topic_note(tag: str, entries: list) -> str:
    entries = sorted(entries, key=lambda e: (e.ts, e.id))
    lines = [
        "---",
        f"topic: {tag}",
        f"entries: {len(entries)}",
        "---",
        "",
        f"# {tag}",
        "",
    ]
    for e in entries:
        preview = (e.text or "").strip().splitlines()
        preview_s = preview[0][:120] if preview else "(no text)"
        lines.append(f"- [[{e.id}]] · {e.type}: {preview_s}")
    return "\n".join(lines).rstrip() + "\n"


def _dream_symbols(entries: list) -> dict[str, int]:
    counts: dict[str, int] = defaultdict(int)
    for e in entries:
        for w in WORD_RE.findall((e.text or "").lower()):
            if w in STOP:
                continue
            counts[w] += 1
    return dict(sorted(counts.items(), key=lambda x: (-x[1], x[0]))[:40])


def _render_dreams_note(dream_entries: list, symbols: dict[str, int]) -> str:
    lines = [
        "---",
        "topic: dreams",
        f"entries: {len(dream_entries)}",
        "---",
        "",
        "# Dreams",
        "",
        "## Symbols",
        "",
    ]
    for sym, n in symbols.items():
        lines.append(f"- {sym} ({n})")
    lines.extend(["", "## Entries", ""])
    for e in sorted(dream_entries, key=lambda x: (x.ts, x.id)):
        preview = (e.text or "").strip().splitlines()
        preview_s = preview[0][:120] if preview else "(voice / empty)"
        lines.append(f"- [[{e.id}]]: {preview_s}")
    return "\n".join(lines).rstrip() + "\n"


def run_topics(
    chronicle_dir: Path | str | None = None,
    *,
    dry_run: bool = False,
) -> dict:
    root = resolve_chronicle_dir(chronicle_dir)
    cfg = ensure_config(root)
    entries = load_all_entries(root, fallback_tz=cfg.timezone)

    by_tag: dict[str, list] = defaultdict(list)
    dreams = []
    for e in entries:
        if e.type == "dream":
            dreams.append(e)
        for t in e.tags:
            if t.startswith("future:") or t.startswith("prompt:"):
                continue
            key = t if t == "#plan" else t.lstrip("#").lower()
            by_tag[key].append(e)

    written: list[str] = []
    for tag, ents in sorted(by_tag.items()):
        slug = topic_slug(tag)
        path = root / "_system" / "derived" / "topics" / f"{slug}.md"
        content = _render_topic_note(tag, ents)
        if write_if_changed(path, content, dry_run=dry_run):
            written.append(str(path))
        if not dry_run and path.is_file():
            mirror_note(path, cfg.vault_mirror)

    if dreams:
        symbols = _dream_symbols(dreams)
        path = root / "_system" / "derived" / "topics" / "dreams.md"
        content = _render_dreams_note(dreams, symbols)
        if write_if_changed(path, content, dry_run=dry_run):
            written.append(str(path))
        if not dry_run and path.is_file():
            mirror_note(path, cfg.vault_mirror)

    log.info("%s %d topic notes", "[dry-run]" if dry_run else "Wrote", len(written))
    return {"written": written, "dreams": len(dreams), "dry_run": dry_run}

File: chronicle-pc/server/src/topics.rs
//! Topic pages + dream clustering (topics.py).

use std::collections::HashMap;
use std::path::{Path, PathBuf};

use regex::Regex;
use serde_json::{json, Value};

use crate::entries as store;
use crate::models::Entry;
use crate::errors::ChronicleError;
use crate::notes::{topic_slug, write_if_changed};
use crate::paths::resolve_chronicle_dir;

const STOPWORDS: [&str; 26] = [
    "that", "this", "with", "from", "have", "were", "been", "they", "them", "then",
    "when", "what", "your", "about", "into", "just", "like", "there", "their",
    "would", "could", "should", "dream", "dreams", "dreamt", "dreamed",
];

pub fn topic_path(root: &Path, tag: &str) -> PathBuf {
    root.join("_system")
        .join("derived")
        .join("topics")
        .join(format!("{}.md", topic_slug(tag)))
}

pub fn run_topics(dir: Option<&Path>, dry_run: bool) -> Result<Value, ChronicleError> {
    let root = resolve_chronicle_dir(dir)?;
    let entries = store::load_all_entries(&root)?;

    // Tag buckets exclude future:/prompt:.
    let mut buckets: HashMap<String, Vec<&Entry>> = HashMap::new();
    for e in &entries {
        for t in &e.tags {
            if t.starts_with("future:") || t.starts_with("prompt:") {
                continue;
            }
            buckets.entry(t.clone()).or_default().push(e);
        }
    }
    let mut keys: Vec<String> = buckets.keys().cloned().collect();
    keys.sort();

    let mut written: Vec<String> = Vec::new();
    for key in &keys {
        let mut list = buckets.remove(key).unwrap_or_default();
        list.sort_by(|a, b| (&a.ts, &a.id).cmp(&(&b.ts, &b.id)));
        let preview = |e: &Entry| -> String {
            let first = e.text.lines().next().unwrap_or("").trim();
            let cut: String = first.chars().take(120).collect();
            if cut.is_empty() { "(no text)".into() } else { cut }
        };
        let mut body = format!(
            "---\ntopic: {}\nentries: {}\n---\n\n# {}\n\n",
            key,
            list.len(),
            key
        );
        for e in &list {
            body.push_str(&format!("- [[{}]] · {}: {}\n", e.id, e.kind, preview(e)));
        }
        let content = format!("{}\n", body.trim_end());
        let path = topic_path(&root, key);
        if write_if_changed(&path, &content, dry_run)? {
            written.push(path.to_string_lossy().to_string());
        }
    }

    // Dream clustering.
    let word_re = Regex::new(r"[a-zA-Z][a-zA-Z0-9']{3,}").unwrap();
    let mut symbols: HashMap<String, i64> = HashMap::new();
    let dreams: Vec<&Entry> = entries.iter().filter(|e| e.kind == "dream").collect();
    for e in &dreams {
        let lower = e.text.to_lowercase();
        for m in word_re.find_iter(&lower) {
            let w = m.as_str();
            if !STOPWORDS.contains(&w) {
                *symbols.entry(w.to_string()).or_default() += 1;
            }
        }
    }
    let mut top: Vec<(String, i64)> = symbols.into_iter().collect();
    top.sort_by(|a, b| (-b.1, &a.0).cmp(&(-a.1, &b.0)));
    top.truncate(40);

    if !dreams.is_empty() {
        let mut body = format!(
            "---\ntopic: dreams\nentries: {}\n---\n\n# Dreams\n\n## Symbols\n\n",
            dreams.len()
        );
        for (sym, n) in &top {
            body.push_str(&format!("- {sym} ({n})\n"));
        }
        body.push_str("\n## Entries\n\n");
        for e in &dreams {
            let first = e.text.lines().next().unwrap_or("").trim();
            let cut: String = first.chars().take(120).collect();
            let preview = if cut.is_empty() { "(voice / empty)".to_string() } else { cut };
            body.push_str(&format!("- [[{}]]: {preview}\n", e.id));
        }
        let content = format!("{}\n", body.trim_end());
        let path = topic_path(&root, "dreams");
        if write_if_changed(&path, &content, dry_run)? {
            written.push(path.to_string_lossy().to_string());
        }
    }
    Ok(json!({ "written": written, "dry_run": dry_run }))
}
